rfd/0145-api-client-credential-reload.md

```diff
@@ -0,0 +1,200 @@
+
+This was initially specified in "RFD 0010 - API and Client Libraries":
+
+> Credential loaders should detect underlying file changes and automatically reload credentials and update the grpc transport.
+
+However, this did not make it into the implementation.
+
+## Why
+
+Teleport encourages the use of short-lived certificates, but consuming these
+with our API client is difficult. 
+
+In the current state, we rely on consumers to implement their own reload 
+mechanism which detects the change to the file and then creates a new client, 
+they then must propagate the new client through their application. This is a 
+relatively challenging feat for less experienced Go engineers and raises the 
+barrier of entry to those who wish to create custom automations.
+
+This not only affects the custom automations, but also our Teleport Access
+Plugins.
+
+This issue has raised in prominence with the growth of Machine ID but as
+Machine ID produces short-lived certificates it is incompatible with usage
+of the API client and Access plugins in their current state.
+
+## Details
+
+It may be helpful to reduce the success criteria of this build to focus on
+short-lived identity files. This simplifies the implementation as an identity
+file contains all relevant material within a single file. This is in contrast
+to a separate key and certificate file where it is possible to reload it in an
+inconsistent state where the key does not match the certificate.
+
+One of the challenges with implementing this is the complexity of transport
+to the Teleport API - we not only need to reload the TLS credentials which are
+presented during the TLS handshake, but also the SSH credentials which are
+used as part of our custom diallers that establish connectivity over tunnels.
+
+We should also aim to support the rotation of Certificate Authorities. Machine
+ID will output rotated CA credentials and having the API client automatically
+load these will improve stability of tools that use the API over CA rotations.
+
+### Option A: Credential reloading
+
+In this option, we build support for reloading files directly into each of the 
+`client.Credential` credential loader implementations. The client itself remains
+unaware that any credential reloading as occurred.
+
+For a developer writing a custom automation, it would look like so:
+
+```go
+func example() {
+	identityFileWatcher, err := client.WatchIdentityFile("my-identity-file")
+	defer identityFileWatcher.Close()
+	
+    clt, err := client.New(ctx, client.Config{
+        Addrs: []string{
+        "tele.example.com:443",
+        },
+        Credentials: []client.Credentials{
+            identityFileWatcher,
+        },
+    })
+	
+	// Or, rather than relying on change detection within the
+	// IdentityFileWatcher, a Reload() could be manually invoked.
+	err = identityFileWatcher.Reload()
+}
+```
+
+The challenge of this is that the `tls.Config` and `ssh.Config` passed into the
```

Comment:
I might be missing something but at least `tls.Config` can dynamically load the certificate for each call if you use the `GetCertificate`.

```
tlsConf:=&tls.Config{
    GetCertificate:  func(*tls.ClientHelloInfo) (*tls.Certificate, error){
       return identityFileWatcher.Load()
    } 
}
```

Reply:
Yeah these are the "callback fields" I mention below. One challenge is that these don't always line up. Specifying a client cert maps one to one between the static and dynamic callback field, but specifying a dynamic CA pool is much harder and forces us to re-implement certificate validation:

````go
		VerifyConnection: func(state tls.ConnectionState) error {
			// This VerifyConnection callback is based on the standard library
			// implementation of verifyServerCertificate in the `tls` package.
			// We provide our own implementation so we can dynamically handle
			// a changing CA Roots pool.
			d.mu.RLock()
			defer d.mu.RUnlock()
			opts := x509.VerifyOptions{
				DNSName:       state.ServerName,
				Intermediates: x509.NewCertPool(),
				Roots:         d.pool,
			}
			for _, cert := range state.PeerCertificates[1:] {
				// Whilst we don't currently use intermediate certs at
				// Teleport, including this here means that we are
				// future-proofed in case we do.
				opts.Intermediates.AddCert(cert)
			}
			_, err := state.PeerCertificates[0].Verify(opts)
			return err
		},
```